[Feat] add at32 dma v1 low level driver

zhaqian12 opened the pull request.

HorrorTroll committed:
Add AT32F415 support

HorrorTroll committed:
Fixed issue of PLLCFGEN not working

HorrorTroll committed:
Merge pull request #2 from qmk-arterytek/at32f415_log

Add AT32F415 basic support

zhaqian12 committed:
[Feat] add at32 dma v1 low level driver

HorrorTroll commented on os/hal/ports/AT32/AT32F41x/platform.mk:
```diff
@@ -28,6 +28,7 @@ include $(CHIBIOS_CONTRIB)/os/hal/ports/AT32/LLD/GPIOv1/driver.mk
 include $(CHIBIOS_CONTRIB)/os/hal/ports/AT32/LLD/OTGv1/driver.mk
 include $(CHIBIOS_CONTRIB)/os/hal/ports/AT32/LLD/SYSTICKv1/driver.mk
 include $(CHIBIOS_CONTRIB)/os/hal/ports/AT32/LLD/TMRv1/driver.mk
+include $(CHIBIOS_CONTRIB)/os/hal/ports/AT32/LLD/DMAv1/driver.mk
```
I think DMA LLD should be above GPIO LLD
```suggestion
include $(CHIBIOS_CONTRIB)/os/hal/ports/AT32/LLD/DMAv1/driver.mk
include $(CHIBIOS_CONTRIB)/os/hal/ports/AT32/LLD/GPIOv1/driver.mk
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+#define DMA2_CH6_VARIANT            0
+#define DMA2_CH7_VARIANT            0
+
+#endif
```
Since DMA_CSELR is only available for STM32F0/L0, so those are unused for AT32F415.
```suggestion
#define DMA1_CH1_VARIANT            0
#define DMA1_CH2_VARIANT            0
#define DMA1_CH3_VARIANT            0
#define DMA1_CH4_VARIANT            0
#define DMA1_CH5_VARIANT            0
#define DMA1_CH6_VARIANT            0
#define DMA1_CH7_VARIANT            0
#define DMA2_CH1_VARIANT            0
#define DMA2_CH2_VARIANT            0
#define DMA2_CH3_VARIANT            0
#define DMA2_CH4_VARIANT            0
#define DMA2_CH5_VARIANT            0
#define DMA2_CH6_VARIANT            0
#define DMA2_CH7_VARIANT            0
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+#if !defined(AT32_DMA1_CH8_CMASK)
+#define AT32_DMA1_CH8_CMASK        (1U << 7U)
+#endif
+
```
Since F415 don't have DMA1 Channel 8 so this goes unused.
```suggestion
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+#if !defined(AT32_DMA2_CH8_CMASK)
+#define AT32_DMA2_CH8_CMASK        (1U << (AT32_DMA1_NUM_CHANNELS + 7U))
+#endif
+
```
Same above
```suggestion
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+#endif
+#if AT32_DMA2_NUM_CHANNELS > 7
+  {DMA2, DMA2_Channel8, AT32_DMA2_CH8_CMASK, DMA2_CH8_VARIANT, 28, 7 + AT32_DMA1_NUM_CHANNELS, AT32_DMA2_CH8_NUMBER},
+#endif
```
Same above
```suggestion
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+   */
+  uint32_t          allocated_mask;
+  /**
+   * @brief   Mask of the enabled streams ISRs.
```
```suggestion
   * @brief   Mask of the enabled streams STSs.
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+  /**
+   * @brief   Mask of the enabled streams ISRs.
+   */
+  uint32_t          isr_mask;
```
```suggestion
  uint32_t          sts_mask;
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+    /**
+     * @brief   DMA callback function.
+     */
+    at32_dmaisr_t    func;
```
```suggestion
    at32_dmasts_t    func;
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+#endif
+
+/*
+ * Default ISR collision masks.
```
```suggestion
 * Default STS collision masks.
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA1 stream 1 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA1 stream 1 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA1 stream 2 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA1 stream 2 shared STS.
 *
 * @sts
```

zhaqian12 replied:
This is the DMA interrupt service routine,but STS is the reg name about dma isr flag.I think ISR is more appropriate, and consistent with at32_isr.c

zhaqian12 replied:
Or something like this 
~~~
/**
 * @brief   DMA1 stream 2 shared interrupt handler.
 *
 * @isr
 */
~~~

HorrorTroll replied:
> Or something like this
> 
> ```
> /**
>  * @brief   DMA1 stream 2 shared interrupt handler.
>  *
>  * @isr
>  */
> ```

Yep, i think this is much more better.

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA1 stream 3 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA1 stream 3 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA1 stream 4 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA1 stream 4 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA1 stream 5 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA1 stream 5 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA1 stream 6 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA1 stream 6 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA1 stream 7 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA1 stream 7 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+  OSAL_IRQ_EPILOGUE();
+}
+#endif
+
```
Unused
```suggestion
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA2 stream 1 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA2 stream 1 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA2 stream 2 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA2 stream 2 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA2 stream 3 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA2 stream 3 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA2 stream 4 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA2 stream 4 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA2 stream 5 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA2 stream 5 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA2 stream 6 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA2 stream 6 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+/**
+ * @brief   DMA2 stream 7 shared ISR.
+ *
+ * @isr
```
```suggestion
 * @brief   DMA2 stream 7 shared STS.
 *
 * @sts
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+  OSAL_IRQ_EPILOGUE();
+}
+#endif
+
```
Unused
```suggestion
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+  dma.allocated_mask = 0U;
+  dma.isr_mask       = 0U;
+  for (i = 0; i < AT32_DMA_STREAMS; i++) {
+    _at32_dma_streams[i].channel->CTRL = AT32_DMA_CCR_RESET_VALUE;
```
```suggestion
  dma.sts_mask       = 0U;
  for (i = 0; i < AT32_DMA_STREAMS; i++) {
    _at32_dma_streams[i].channel->CTRL = AT32_DMA_CTRL_RESET_VALUE;
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+ */
+const at32_dma_stream_t *dmaStreamAllocI(uint32_t id,
+                                          uint32_t priority,
+                                          at32_dmaisr_t func,
```
```suggestion
                                          at32_dmasts_t func,
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+
+      /* Putting the stream in a known state.*/
+      dmaStreamDisable(dmastp);
+      dmastp->channel->CTRL = AT32_DMA_CCR_RESET_VALUE;
```
```suggestion
        if ((dma.sts_mask & dmastp->cmask) == 0U) {
          nvicEnableVector(dmastp->vector, priority);
        }
        dma.sts_mask |= mask;
      }

      /* Putting the stream in a known state.*/
      dmaStreamDisable(dmastp);
      dmastp->channel->CTRL = AT32_DMA_CTRL_RESET_VALUE;
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+ */
+const at32_dma_stream_t *dmaStreamAlloc(uint32_t id,
+                                         uint32_t priority,
+                                         at32_dmaisr_t func,
```
```suggestion
                                         at32_dmasts_t func,
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+  dma.isr_mask &= ~(1U << selfindex);
+
+  /* Disables the associated IRQ vector if it is no more in use.*/
+  if ((dma.isr_mask & dmastp->cmask) == 0U) {
```
```suggestion
  dma.sts_mask &= ~(1U << selfindex);

  /* Disables the associated IRQ vector if it is no more in use.*/
  if ((dma.sts_mask & dmastp->cmask) == 0U) {
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+  uint32_t flags;
+  uint32_t selfindex = (uint32_t)dmastp->selfindex;
+
+  flags = (dmastp->dma->STS >> dmastp->shift) & AT32_DMA_ISR_MASK;
```
```suggestion
  flags = (dmastp->dma->STS >> dmastp->shift) & AT32_DMA_STS_MASK;
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,796 @@
+#if (AT32_DMA_SUPPORTS_DMAMUX == TRUE) || defined(__DOXYGEN__)
+/**
+ * @brief   Associates a peripheral request to a DMA stream.
+ * @note    This function can be invoked in both ISR or thread context.
```
```suggestion
 * @note    This function can be invoked in both STS or thread context.
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+/**
+ * @brief   Mask of the ISR bits passed to the DMA callback functions.
+ */
+#define AT32_DMA_ISR_MASK          0x0E
```
```suggestion
 * @brief   Mask of the STS bits passed to the DMA callback functions.
 */
#define AT32_DMA_STS_MASK          0x0E
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+#endif
+#if AT32_DMA1_NUM_CHANNELS > 7
+#define AT32_DMA1_STREAM8          AT32_DMA_STREAM(7)
+#endif
```
Unused
```suggestion
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+#endif
+#if AT32_DMA2_NUM_CHANNELS > 7
+#define AT32_DMA2_STREAM8          AT32_DMA_STREAM(AT32_DMA1_NUM_CHANNELS + 7)
+#endif
```
Unused
```suggestion
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+#define AT32_DMA_CR_SIZE_MASK      (AT32_DMA_CR_PSIZE_MASK |              \
+                                     AT32_DMA_CR_MSIZE_MASK)
+#define AT32_DMA_CR_PL_MASK        DMA_CTRL_CHPL
+#define AT32_DMA_CR_PL(n)          ((n) << 12U)
```
```suggestion
 * @name    CCTRL register constants common to all DMA types
 * @{
 */
#define AT32_DMA_CCTRL_RESET_VALUE  0x00000000U
#define AT32_DMA_CCTRL_CHEN         DMA_CCTRL_CHEN
#define AT32_DMA_CCTRL_DTERRIEN     DMA_CCTRL_DTERRIEN
#define AT32_DMA_CCTRL_HDTIEN       DMA_CCTRL_HDTIEN
#define AT32_DMA_CCTRL_FDTIEN       DMA_CCTRL_FDTIEN
#define AT32_DMA_CCTRL_DTD_MASK     (DMA_CCTRL_DTD | DMA_CCTRL_M2M)
#define AT32_DMA_CCTRL_DTD_P2M      0U
#define AT32_DMA_CCTRL_DTD_M2P      DMA_CCTRL_DTD
#define AT32_DMA_CCTRL_DTD_M2M      DMA_CCTRL_M2M
#define AT32_DMA_CCTRL_LM           DMA_CCTRL_LM
#define AT32_DMA_CCTRL_PINCM        DMA_CCTRL_PINCM
#define AT32_DMA_CCTRL_MINCM        DMA_CCTRL_MINCM
#define AT32_DMA_CCTRL_PWIDTH_MASK  DMA_CCTRL_PWIDTH
#define AT32_DMA_CCTRL_PWIDTH_BYTE  0U
#define AT32_DMA_CCTRL_PWIDTH_HWORD DMA_CCTRL_PWIDTH_0
#define AT32_DMA_CCTRL_PWIDTH_WORD  DMA_CCTRL_PWIDTH_1
#define AT32_DMA_CCTRL_MWIDTH_MASK  DMA_CCTRL_MWIDTH
#define AT32_DMA_CCTRL_MWIDTH_BYTE  0U
#define AT32_DMA_CCTRL_MWIDTH_HWORD DMA_CCTRL_MWIDTH_0
#define AT32_DMA_CCTRL_MWIDTH_WORD  DMA_CCTRL_MWIDTH_1
#define AT32_DMA_CCTRL_SIZE_MASK    (AT32_DMA_CCTRL_PWIDTH_MASK |           \
                                     AT32_DMA_CCTRL_MWIDTH_MASK)
#define AT32_DMA_CCTRL_CHPL_MASK    DMA_CCTRL_CHPL
#define AT32_DMA_CCTRL_CHPL(n)      ((n) << 12U)
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+#define AT32_DMA_ISR_HTIF          (0x1U << 2)
+#define AT32_DMA_ISR_TCIF          (0x1U << 1)
+/** @} */
+
```
Unused
```suggestion
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+#if AT32_DMA_SUPPORTS_DMAMUX && AT32_DMA_SUPPORTS_CSELR
+#error "AT32_DMA_SUPPORTS_DMAMUX and AT32_DMA_SUPPORTS_CSELR both TRUE"
+#endif
+
```
CSELR not supported
```suggestion
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+ * @param[in] flags     pre-shifted content of the ISR register, the bits
+ *                      are aligned to bit zero
+ */
+typedef void (*at32_dmaisr_t)(void *p, uint32_t flags);
```
```suggestion
 * @param[in] flags     pre-shifted content of the STS register, the bits
 *                      are aligned to bit zero
 */
typedef void (*at32_dmasts_t)(void *p, uint32_t flags);
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+  uint8_t               dummy;          /**< @brief Filler.                 */
+#endif
+  uint8_t               shift;          /**< @brief Bit offset in ISR, CLR
+                                             and CSELR registers.           */
```
```suggestion
  uint8_t               dummy;          /**< @brief Filler.                 */
  uint8_t               shift;          /**< @brief Bit offset in STS, CLR  */
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+#define dmaStreamSetMode(dmastp, mode) {                                    \
+  (dmastp)->channel->CTRL  = (uint32_t)(mode);                               \
+}
+#endif
```
```suggestion
 * @param[in] mode      value to be written in the CCTRL register
 *
 * @special
 */
#define dmaStreamSetMode(dmastp, mode) {                                    \
  (dmastp)->channel->CCTRL = (uint32_t)(mode);                              \
}
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+ * @special
+ */
+#define dmaStreamEnable(dmastp) {                                           \
+  (dmastp)->channel->CTRL |= AT32_DMA_CR_EN;                                \
```
```suggestion
  (dmastp)->channel->CCTRL |= AT32_DMA_CCTRL_CHEN;                          \
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+ */
+#define dmaStreamDisable(dmastp) {                                          \
+  (dmastp)->channel->CTRL &= ~(AT32_DMA_CR_TCIE | AT32_DMA_CR_HTIE |       \
+                              AT32_DMA_CR_TEIE | AT32_DMA_CR_EN);         \
```
```suggestion
  (dmastp)->channel->CCTRL &= ~(AT32_DMA_CCTRL_FDTIEN |                     \
                                AT32_DMA_CCTRL_HDTIEN |                     \
                                AT32_DMA_CCTRL_DTERRIEN |                   \
                                AT32_DMA_CCTRL_CHEN);                       \
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+ * @special
+ */
+#define dmaStreamClearInterrupt(dmastp) {                                   \
+  (dmastp)->dma->CLR = AT32_DMA_ISR_MASK << (dmastp)->shift;              \
```
```suggestion
  (dmastp)->dma->CLR = AT32_DMA_STS_MASK << (dmastp)->shift;              \
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+ * @post    After use the stream can be released using @p dmaStreamRelease().
+ *
+ * @param[in] dmastp    pointer to a at32_dma_stream_t structure
+ * @param[in] mode      value to be written in the CCR register, this value
+ *                      is implicitly ORed with:
+ *                      - @p AT32_DMA_CR_MINC
+ *                      - @p AT32_DMA_CR_PINC
+ *                      - @p AT32_DMA_CR_DIR_M2M
+ *                      - @p AT32_DMA_CR_EN
```
```suggestion
 * @param[in] mode      value to be written in the CCTRL register, this value
 *                      is implicitly ORed with:
 *                      - @p AT32_DMA_CCTRL_MINCM
 *                      - @p AT32_DMA_CCTRL_PINCM
 *                      - @p AT32_DMA_CCTRL_DTD_M2M
 *                      - @p AT32_DMA_CCTRL_CHEN
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+  dmaStreamSetMemory0(dmastp, dst);                                         \
+  dmaStreamSetTransactionSize(dmastp, n);                                   \
+  dmaStreamSetMode(dmastp, (mode) |                                         \
+                           AT32_DMA_CR_MINC | AT32_DMA_CR_PINC |          \
+                           AT32_DMA_CR_DIR_M2M | AT32_DMA_CR_EN);         \
```
```suggestion
                           AT32_DMA_CCTRL_MINCM | AT32_DMA_CCTRL_PINCM |    \
                           AT32_DMA_CCTRL_DTD_M2M | AT32_DMA_CCTRL_CHEN);   \
```

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,556 @@
+  void dmaInit(void);
+  const at32_dma_stream_t *dmaStreamAllocI(uint32_t id,
+                                            uint32_t priority,
+                                            at32_dmaisr_t func,
+                                            void *param);
+  const at32_dma_stream_t *dmaStreamAlloc(uint32_t id,
+                                           uint32_t priority,
+                                           at32_dmaisr_t func,
```
```suggestion
                                            at32_dmasts_t func,
                                            void *param);
  const at32_dma_stream_t *dmaStreamAlloc(uint32_t id,
                                           uint32_t priority,
                                           at32_dmasts_t func,
```

HorrorTroll reviewed (changes requested):
Thanks for your pr, I've just take a look at reviewing your code. There are some code need to be changed first. And it seem there are 2 folder DMAv1 on LLD

zhaqian12 committed:
Apply suggestions from code review

Apply suggestions from code review

Update os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c

Update os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h

Update os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c

Update os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h

Update os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h

Update os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h

Update os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.h

Update os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c

Update os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c

Update os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c

Update os/hal/ports/AT32/AT32F41x/platform.mk

Co-Authored-By: HorrorTroll <sonicvipduc@gmail.com>

zhaqian12 committed:
update at32 dmav1 lld

zhaqian12 commented:
Thanks for you review and suggestion and sorry for my carelessness. The changes have been committed, please check again, thank you!

HorrorTroll commented:
It fine, I'll re-take a look after I've got some free time 👍. Thanks you for your hard work

zhaqian12 commented:
Sorry, please wait a minute, some bugs are not fixed and commited yet

zhaqian12 committed:
fix typo in at32_dma

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/at32_dma.c:
```diff
@@ -0,0 +1,718 @@
+#endif
+#if AT32_DMA1_NUM_CHANNELS > 7
+  {DMA1, DMA1_Channel8, AT32_DMA1_CH8_CMASK, DMA1_CH8_VARIANT, 28, 7, AT32_DMA1_CH8_NUMBER},
+#endif
```
I've forgot this on the last reviewer
```suggestion
```

zhaqian12 committed:
fix at32 dma isr

HorrorTroll commented on os/hal/ports/AT32/LLD/DMAv1/at32_dma.h:
```diff
@@ -0,0 +1,498 @@
+#define AT32_DMA_CCTRL_CHPL_MASK    DMA_CCTRL_CHPL
+#define AT32_DMA_CCTRL_CHPL(n)      ((n) << 12U)
+/** @} */
+
```
Forgot about this, it needed for I2C and other peripherals.
```suggestion

/**
 * @name    Status flags passed to the ISR callbacks
 * @{
 */
#define AT32_DMA_STS_DTERRF    (0x1U << 3)
#define AT32_DMA_STS_HDTF        (0x1U << 2)
#define AT32_DMA_STS_FDTF         (0x1U << 1)
/** @} */

```

HorrorTroll reviewed (commented):
Last 2 need to fix and should be done. Make sure to testing out first both flexible and solid DMA on QMK and see if it work or not.

zhaqian12 committed:
Apply suggestions from code review

Co-authored-by: HorrorTroll <sonicvipduc@gmail.com>

zhaqian12 commented:
Thanks a lot for your support!

HorrorTroll committed:
Merge branch 'arterytek_merge' into at32_dma
